sound/soc/intel/boards/sof_maxim_common.h

```diff
@@ -0,0 +1,87 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright(c) 2020 Intel Corporation.
+ */
+
+/*
+ * This file defines data structures used in Machine Driver for Intel
+ * platforms with Maxim Codecs.
+ */
+#ifndef __SOF_MAXIM_COMMON_H
+#define __SOF_MAXIM_COMMON_H
+
+#define MAX98373_CODEC_DAI	"max98373-aif1"
+#define MAX98373_DEV0_NAME	"i2c-MX98373:00"
+#define MAX98373_DEV1_NAME	"i2c-MX98373:01"
+
+static int max98373_hw_params(struct snd_pcm_substream *substream,
+			      struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai;
+	int j;
+
+	for_each_rtd_codec_dai(rtd, j, codec_dai) {
+		if (!strcmp(codec_dai->component->name, MAX98373_DEV0_NAME)) {
+			/* DEV0 tdm slot configuration */
+			snd_soc_dai_set_tdm_slot(codec_dai, 0x30, 3, 8, 16);
+		}
+		if (!strcmp(codec_dai->component->name, MAX98373_DEV1_NAME)) {
+			/* DEV1 tdm slot configuration */
+			snd_soc_dai_set_tdm_slot(codec_dai, 0xC0, 3, 8, 16);
+		}
+	}
+	return 0;
+}
+
+static struct snd_soc_ops max98373_ops = {
+	.hw_params = max98373_hw_params,
+};
+
+static struct snd_soc_codec_conf max98373_codec_conf[] = {
+	{
+		.dlc = COMP_CODEC_CONF(MAX98373_DEV0_NAME),
+		.name_prefix = "Right",
+	},
+
+	{
+		.dlc = COMP_CODEC_CONF(MAX98373_DEV1_NAME),
+		.name_prefix = "Left",
+	},
+};
+
+static struct snd_soc_dai_link_component max98373_components[] = {
+	{  /* For Left */
+		.name = MAX98373_DEV0_NAME,
+		.dai_name = MAX98373_CODEC_DAI,
+	},
+	{  /* For Right */
+		.name = MAX98373_DEV1_NAME,
+		.dai_name = MAX98373_CODEC_DAI,
+	},
+};
+
+static const struct snd_soc_dapm_route max98373_dapm_routes[] = {
+	/* speaker */
+	{ "Left Spk", NULL, "Left BE_OUT" },
+	{ "Right Spk", NULL, "Right BE_OUT" },
+};
```

Comment:
> This doesn't feel quite right... Defining static data in a header means every other file `.c` or `.h` that includes this header will have identical instances of these objects... I understand that might indeed be what you want, because the goal of this header is to only be included by files, that need those instances, but this is somewhat unsafe (it's easy to overlook them and include the header in another file just to get struct defines or similar. I think a better idea would be to define macros here for these objects and to then instantiate them in each `.c` file separately

ok, I can move the function definitions to sof_maxim_common.c and will include sof_maxim_common.o in Makefile for the needed one. In the header file, let me remove static.
Does this method ok?

Reply:
That's a valid point @lyakh We don't have any good examples in our boards file now. 

In skl_hda_dsp_common.[hc] I used following:

header:
extern struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS];

c-file for common part:
struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS] = { // actual definition }

... but this a bit annoying if you don't need the implementation file for anything else.

Maybe macros would be cleaner still? Any good examples in kernel tree @lyakh ?

Reply:
yes we can use modules, but we'd want to use namespaces so that the maxim_common symbols are only used from specific machine drivers.

Reply:
@kv2019i examples depend of how exactly those objects should be used. From this implementation it looks like every user would have exactly the same object definitions - same names, same contents. But some of them aren't `const`. Does that mean that they really can be modified at run-time? I just checked, it wouldn't be possible to make at least some of them `const`. Still, if they aren't really going to change, they can be placed in a common .c. If they can indeed change and every user should get a private copy of them, then we could use macros similar to those in soc-dapm.h.

Reply:
>but this is somewhat unsafe (it's easy to overlook them and include the header in another file just to get struct defines or similar.
Hi @lyakh ,
I have now updated in new patchset ..only keeping needed in header file and moved others to sof_maxim_common.c